src/Stratis.Bitcoin/EventBus/InMemoryEventBus.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Stratis.Bitcoin.EventBus
+{
+    public class InMemoryEventBus : IEventBus
+    {
+        /// <summary>
+        /// The subscriptions stored by EventType
+        /// </summary>
+        private readonly Dictionary<Type, List<ISubscription>> subscriptions;
+
+        /// <summary>
+        /// The subscriptions lock to prevent race condition during publishing
+        /// </summary>
+        private static readonly object SubscriptionsLock = new object();
+
+        public InMemoryEventBus()
+        {
+            this.subscriptions = new Dictionary<Type, List<ISubscription>>();
+        }
+
+        /// <inheritdoc />
+        public SubscriptionToken Subscribe<TEventBase>(Action<TEventBase> action) where TEventBase : EventBase
+        {
+            if (action == null)
+                throw new ArgumentNullException(nameof(action));
+
+            lock (SubscriptionsLock)
+            {
+                if (!this.subscriptions.ContainsKey(typeof(TEventBase)))
+                    this.subscriptions.Add(typeof(TEventBase), new List<ISubscription>());
+
+                var token = new SubscriptionToken(this, typeof(TEventBase));
+                this.subscriptions[typeof(TEventBase)].Add(new Subscription<TEventBase>(action, token));
+                return token;
+            }
+        }
+
+        /// <inheritdoc />
+        public void Unsubscribe(SubscriptionToken token)
+        {
+            if (token == null)
+                throw new ArgumentNullException(nameof(token));
+
+            lock (SubscriptionsLock)
+            {
+                if (this.subscriptions.ContainsKey(token.EventType))
+                {
+                    var allSubscriptions = this.subscriptions[token.EventType];
+
+                    var subscriptionToRemove = allSubscriptions.FirstOrDefault(sub => sub.SubscriptionToken.Token == token.Token);
+                    if (subscriptionToRemove != null)
+                        this.subscriptions[token.EventType].Remove(subscriptionToRemove);
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public void Publish<TEventBase>(TEventBase eventItem) where TEventBase : EventBase
+        {
+            if (eventItem == null)
+                throw new ArgumentNullException(nameof(eventItem));
+
+            // Assigns an unique id to the event.
+            eventItem.CorrelationId = Guid.NewGuid();
+
+            List<ISubscription> allSubscriptions = new List<ISubscription>();
+            lock (SubscriptionsLock)
+            {
+                if (this.subscriptions.ContainsKey(typeof(TEventBase)))
+                    allSubscriptions = this.subscriptions[typeof(TEventBase)];
+            }
+
+            for (var index = 0; index < allSubscriptions.Count; index++)
+            {
+                var subscription = allSubscriptions[index];
```

Comment:
I think there's a possible index out of bounds here. Consider the case where we are enumerating through `allSubscriptions` and the subscription at the current `index` has unsubscribed since `allSubscriptions` was assigned.

Reply:
It shouldn't get an out of bounds, `allSubscriptions` is a local variable created above. However, you would get an exception if you are trying to broadcast to a subscription that happened to be disposed or removed, but I guess this is handled in the catch.

Reply:
no there can't be an index out of bound because allSubscriptions is a loca list built in a lock (the same used on unsubscribe)

Reply:
The lock is taken while `allSubscriptions` is assigned but released before it is accessed. So it is still possible for `allSubscriptions` to have elements added/removed in `Unsubscribe` because the lock is no longer taken.

Reply:
ah no wait I see what you mean, I don't know the better way between:
- a ToList of `this.subscriptions[typeof(TEventBase)]` 
- a wider lock
- a different lock object by type (dictionary of locksable objects)

I'd skip the 2nd (a wider lock) because it will hold a lock for too much time.

Reply:
probably I'd just use the ToList for simplicity, we are always in time for improvement, I don't expect this implementation be the definitive one

Reply:
I used ToList

Reply:
`ToList` should work, though it's still possible that subscriptions could receive an event after unsubscribing. Might affect the GCing of unsubscribed objects too, but maybe we don't care about that yet.